libraries/lib-snapping/SnapUtils.cpp

```diff
@@ -0,0 +1,316 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  SnapUtils.cpp
+
+  Dmitry Vedenko
+
+**********************************************************************/
+
+#include "SnapUtils.h"
+
+#include <algorithm>
+#include <cmath>
+#include <unordered_map>
+
+namespace
+{
+const wxString SnapModeKey = L"/Snap/Mode";
+const wxString SnapToKey = L"/Snap/To";
+const wxString OldSnapToKey = L"/SnapTo";
+const auto PathStart = L"SnapFunctions";
+
+struct RegistryVisitor : public Registry::Visitor
+{
+   explicit RegistryVisitor(SnapRegistryVisitor& _visitor)
+       : visitor(_visitor)
+   {
+   }
+   
+   void BeginGroup(Registry::GroupItem& item, const Path&) override
+   {
+      auto group = dynamic_cast<SnapRegistryGroup*>(&item);
+
+      if (group != nullptr)
+         visitor.BeginGroup(*group);
+   }
+
+   void EndGroup(Registry::GroupItem& item, const Path&) override
+   {
+      auto group = dynamic_cast<SnapRegistryGroup*>(&item);
+
+      if (group != nullptr)
+         visitor.EndGroup(*group);
+   }
+
+   void Visit(Registry::SingleItem& item, const Path&) override
+   {
+      auto concreteItem = dynamic_cast<SnapRegistryItem*>(&item);
+
+      if (concreteItem != nullptr)
+         visitor.Visit(*concreteItem);
+   }
+
+   SnapRegistryVisitor& visitor;
+};
+} // namespace
+
+StringSetting SnapToSetting { SnapToKey, "seconds" };
+
+EnumSetting<SnapMode> SnapModeSetting {
+   SnapModeKey,
+   EnumValueSymbols { L"OFF", L"NEAREST", L"PRIOR" },
+   0,
+   { SnapMode::SNAP_OFF, SnapMode::SNAP_NEAREST, SnapMode::SNAP_PRIOR }
+};
+
+SnapMode ReadSnapMode()
+{
+   if (gPrefs->HasEntry(SnapModeKey))
+      return SnapModeSetting.ReadEnum();
+
+   return static_cast<SnapMode>(gPrefs->Read(OldSnapToKey, 0L));
+}
+
+Identifier ReadSnapTo()
+{
+   if (gPrefs->HasEntry(SnapToKey))
+      return SnapToSetting.Read();
+
+   return {};
+}
+
+Registry::GroupItem& SnapFunctionsRegistry::Registry()
+{
+   static Registry::TransparentGroupItem<> registry { PathStart };
+   return registry;
+}
+
+void SnapFunctionsRegistry::Visit(SnapRegistryVisitor& visitor)
+{
+   static Registry::OrderingPreferenceInitializer init {
+      PathStart,
+      { { L"", L"beats,triplets,time,video,cd" } },
```

Comment:
All of which are defined below.  So, why a registry?  Do you anticipate more, and scattered add-on cases?

Reply:
The whole registry thing was done in anticipation of your criticism about the lack of open endlessness.  I would personally agree with this criticism this time, so I took an open ended approach. 

Doing it not open-endedly or even writing "own" would be much quicker, however I have tried to stick to the standard approach used in other places in Audacity. This has costed me quite a lot of time.

Reply:
What did not become a registry yet is the table of formats in NumericConverter.  I thought I might see that, when you mentioned registries.

I hope you agree that registries have their value and will be easier for you to use again.